File: frontend/src/api/developer-agent.ts
/**
 * Developer Agent API - PRD P4-001
 *
 * The Developer Agent takes a requirement (US work item + optional design doc),
 * generates code, commits it to a Git provider (GitHub), and opens a Pull
 * Request. Each run is tracked as a DeveloperJob with a full lifecycle:
 * pending → analyzing → generating → committing → opening_pr → completed.
 */
import api from './index'

export type DeveloperJobStatus =
  | 'pending'
  | 'analyzing'
  | 'generating'
  | 'committing'
  | 'opening_pr'
  | 'completed'
  | 'failed'
  | 'cancelled'

export interface GeneratedFile {
  path: string
  content: string
  mode?: string
}

export interface DeveloperJob {
  id: number
  workspace_id: number
  project_id?: number
  issue_id?: number
  agent_task_id?: number
  git_provider: string
  git_connection_id?: number
  title: string
  requirement_text: string
  design_doc_url?: string
  input_context: Record<string, unknown>
  branch_name: string
  base_branch: string
  commit_message: string
  generated_files: GeneratedFile[]
  pr_number?: number
  pr_url?: string
  pr_title?: string
  commit_sha?: string
  status: DeveloperJobStatus
  progress: number
  current_step?: string
  error_message?: string
  started_at?: string
  completed_at?: string
  cancelled_at?: string
  created_at: string
  updated_at: string
}

export interface DeveloperJobCreate {
  title: string
  requirement_text?: string
  design_doc_url?: string
  project_id?: number
  issue_id?: number
  git_connection_id: number
  git_provider?: string
  branch_name?: string
  base_branch?: string
  commit_message?: string
  pr_title?: string
  pr_body?: string
  language?: string
  files?: GeneratedFile[]
  input_context?: Record<string, unknown>
}

export interface DeveloperJobListFilter {
  status?: DeveloperJobStatus
  limit?: number
}

export const developerAgentApi = {
  list(workspaceId: number, filter: DeveloperJobListFilter = {}): Promise<DeveloperJob[]> {
    const params: Record<string, string | number> = {}
    if (filter.status) params.status = filter.status
    if (filter.limit) params.limit = filter.limit
    return api
      .get(`/workspaces/${workspaceId}/developer-agent/jobs`, { params })
      .then(res => res.data)
  },

  get(workspaceId: number, jobId: number): Promise<DeveloperJob> {
    return api
      .get(`/workspaces/${workspaceId}/developer-agent/jobs/${jobId}`)
      .then(res => res.data)
  },

  create(workspaceId: number, payload: DeveloperJobCreate): Promise<DeveloperJob> {
    return api
      .post(`/workspaces/${workspaceId}/developer-agent/jobs`, payload)
      .then(res => res.data)
  },

  cancel(workspaceId: number, jobId: number): Promise<DeveloperJob> {
    return api
      .post(`/workspaces/${workspaceId}/developer-agent/jobs/${jobId}/cancel`)
      .then(res => res.data)
  },

  delete(workspaceId: number, jobId: number): Promise<void> {
    return api
      .delete(`/workspaces/${workspaceId}/developer-agent/jobs/${jobId}`)
      .then(() => undefined)
  }
}

File: backend/internal/model/developer_agent.go
package model

import (
	"encoding/json"
	"time"
)

// DeveloperJobStatus defines the lifecycle state of a Developer Agent job.
//
// Flow: pending → analyzing → generating → committing → opening_pr → completed
// Any step may transition to "failed".
type DeveloperJobStatus string

const (
	DeveloperJobPending    DeveloperJobStatus = "pending"
	DeveloperJobAnalyzing  DeveloperJobStatus = "analyzing"
	DeveloperJobGenerating DeveloperJobStatus = "generating"
	DeveloperJobCommitting DeveloperJobStatus = "committing"
	DeveloperJobOpeningPR  DeveloperJobStatus = "opening_pr"
	DeveloperJobCompleted  DeveloperJobStatus = "completed"
	DeveloperJobFailed     DeveloperJobStatus = "failed"
	DeveloperJobCancelled  DeveloperJobStatus = "cancelled"
)

// DeveloperJob tracks a single Developer Agent execution (PRD P4-001).
//
// The Developer Agent analyses a requirement (issue + design doc), generates
// source code, commits it to a Git provider, and opens a Pull Request. Each
// job records the inputs, intermediate artifacts, and final PR reference so
// the lifecycle can be observed and audited from the UI.
type DeveloperJob struct {
	BaseModel

	WorkspaceID uint64  `gorm:"not null;index" json:"workspace_id"`
	ProjectID   *uint64 `gorm:"index" json:"project_id,omitempty"`
	IssueID     *uint64 `gorm:"index" json:"issue_id,omitempty"`
	AgentTaskID *uint64 `gorm:"index" json:"agent_task_id,omitempty"`

	// Git provider connection (GitHubConnection.ID, GitIntegration.ID, etc.)
	GitProvider     string  `gorm:"size:20;default:github" json:"git_provider"`
	GitConnectionID *uint64 `gorm:"index" json:"git_connection_id,omitempty"`

	// Inputs
	Title           string          `gorm:"size:255;not null" json:"title"`
	RequirementText string          `gorm:"type:text" json:"requirement_text"`
	DesignDocURL    *string         `gorm:"size:500" json:"design_doc_url,omitempty"`
	InputContext    json.RawMessage `gorm:"type:jsonb;default:'{}'" json:"input_context"`

	// Branch + commit metadata
	BranchName    string `gorm:"size:255" json:"branch_name"`
	BaseBranch    string `gorm:"size:255;default:main" json:"base_branch"`
	CommitMessage string `gorm:"type:text" json:"commit_message"`

	// Generated code artifacts: array of {path, content, mode}
	GeneratedFiles json.RawMessage `gorm:"type:jsonb;default:'[]'" json:"generated_files"`

	// PR reference returned by the git provider
	PRNumber  *int    `gorm:"index" json:"pr_number,omitempty"`
	PRURL     *string `gorm:"size:500" json:"pr_url,omitempty"`
	PRTitle   *string `gorm:"size:255" json:"pr_title,omitempty"`
	CommitSHA *string `gorm:"size:64" json:"commit_sha,omitempty"`

	// Lifecycle
	Status       DeveloperJobStatus `gorm:"size:20;default:pending;index" json:"status"`
	Progress     int                `gorm:"default:0" json:"progress"`
	CurrentStep  *string            `gorm:"size:100" json:"current_step,omitempty"`
	ErrorMessage *string            `gorm:"type:text" json:"error_message,omitempty"`

	StartedAt   *time.Time `json:"started_at,omitempty"`
	CompletedAt *time.Time `json:"completed_at,omitempty"`
	CancelledAt *time.Time `json:"cancelled_at,omitempty"`

	// Relationships
	Workspace Workspace `gorm:"foreignKey:WorkspaceID" json:"-"`
	Project   *Project  `gorm:"foreignKey:ProjectID" json:"-"`
	Issue     *Issue    `gorm:"foreignKey:IssueID" json:"-"`
}

func (DeveloperJob) TableName() string { return "developer_jobs" }
